cpp/src/io/parquet/writer_impl.cu

```diff
@@ -643,20 +683,35 @@ void writer::impl::write_chunk(table_view const &table, pq_chunked_state &state)
   // we actually have.
   std::vector<parquet_column_view> parquet_columns;
   parquet_columns.reserve(num_columns);  // Avoids unnecessary re-allocation
+
+  // because the repetition type is global (in the sense of, not per-rowgroup or per write_chunk()
+  // call) we cannot know up front if the user is going to end up passing tables with nulls/no nulls
+  // in the multiple write_chunk() case.  so we'll do some special handling.
+  // The user can pass in information about the nullability of a column to be enforced across
+  // write_chunk() calls, in a flattened bool vector. Figure out that per column.
+  auto per_column_nullability =
+    (state.single_write_mode)
+      ? std::vector<std::vector<bool>>{}
+      : get_per_column_nullability(table, state.user_metadata_with_nullability.column_nullable);
+
   for (auto it = table.begin(); it < table.end(); ++it) {
     const auto col        = *it;
     const auto current_id = parquet_columns.size();
 
     num_rows = std::max<uint32_t>(num_rows, col.size());
-    parquet_columns.emplace_back(
-      current_id, col, state.user_metadata, state.int96_timestamps, state.stream);
-  }
 
-  if (state.user_metadata_with_nullability.column_nullable.size() > 0) {
-    CUDF_EXPECTS(state.user_metadata_with_nullability.column_nullable.size() ==
-                   static_cast<size_t>(num_columns),
-                 "When passing values in user_metadata_with_nullability, data for all columns must "
-                 "be specified");
+    // if the user is explicitly saying "I am only calling this once", assume the columns in this
```

Comment:
How does a user say that?

Reply:
By setting `state.single_write_mode = true` which is checked in the line below.